src/_guides/libraries/c-interop.md

```diff
@@ -0,0 +1,142 @@
+| [pubspec.yaml]({{ page.hw}}/pubspec.yaml) | The usual Dart [pubspec](/tools/pub/pubspec), with a lower bounds on the SDK that's at least 2.5. |
+| [c/hello.h]({{ page.hw}}/c/hello.h) | Declares the `hello_world()` function. |
+| [c/hello.c]({{ page.hw}}/c/hello.c) | A C file that imports `hello.h` and defines the `hello_world()` function. |
+| [c/Makefile]({{ page.hw}}/c/Makefile) | A macOS-specific build file that compiles the C code into a dynamic library. |
+{:.table .table-striped }
+
+[PENDING: say something about setup.sh? It doesn't seem necessary for this example, but maybe it's needed by other examples?]
+<!-- 
+  | [setup.sh]({{ page.hw}}/setup.sh) | A macOS-specific script that sets an environment variable. [PENDING: Omit from this list? Why is it necessary? I didn't seem to need it.] |
+-->
+
+Building the C library creates two additional files:
+
+| **Generated file** | **Description** |
+| hello_world.dylib | The dynamic library loaded by the Dart app. |
+| c/hello.o | An intermediate object file. |
+{:.table .table-striped }
+
+### Building and running
+
+Here's an example of building the dynamic library and executing the Dart app:
+
+```terminal
+$ cd c
+$ make dylib
+gcc -dynamiclib -undefined suppress -flat_namespace hello.o -o ../hello_world.dylib
+$ cd ..
+$ dart hello.dart
+Hello World
+$ 
+```
+
+### Using dart:ffi
+
+The [`hello.dart` file]({{ page.hw}}/hello.dart)
+illustrates the steps for using dart:ffi to call a C function:
+
+1. Import dart:ffi.
+2. Create a typedef with the FFI type signature of the C function.
+3. Create a typedef for the variable that you'll use when calling the C function.
+4. Open the dynamic library that contains the C function.
+5. Get a reference to the C function, and put it into a variable.
+6. Call the C function.
+
+Here's the code for each step.
+
+1. Import dart:ffi.
+```dart
+import 'dart:ffi' as ffi;
+```
+
+2. Create a typedef with the FFI type signature of the C function. <br>
+   Commonly used types defined by dart:ffi library include
+   `Double`, `Int32`, `NativeFunction`, `Pointer`, `Struct`, `Uint8`, and `Void`.
+```dart
+typedef hello_world_func = ffi.Void Function();
+```
+
+3. Create a typedef for the variable that you'll use when calling the C function.
+```dart
+typedef HelloWorld = void Function();
+```
+
+4. Open the dynamic library that contains the C function.
+```dart
+  final dylib = ffi.DynamicLibrary.open('hello_world.dylib');
+```
+
+5. Get a reference to the C function, and put it into a variable.
+   This code uses the typedefs defined in steps 2 and 3, along with
+   the dynamic library variable from step 4.
+```dart
+  final HelloWorld hello = dylib
+      .lookup<ffi.NativeFunction<hello_world_func>>('hello_world')
+      .asFunction();
+```
+
+6. Call the C function.
+```dart
+  hello();
+```
+
+Once you understand the hello_world example, you should be ready to look at the
+[other dart:ffi examples](#examples).
+
+
+## Loading and packaging C libraries
```

Comment:
Can we swap these? Logically you need to package the C lib first so that you can load it later